Machine-maintenance operator proposal (#341)
openshift/enhancements: enhancements/maintenance/machine-maintenance-operator.md

- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Open Questions
>1. How applicable is this use-case to baremetal installations? 

>2. The ability to declare maintenance windows is currently being discussed in OSD. Nothing has be design in concrete meaning that this proposal has an undesigned depedency. However develop could commence on this operator and have the maintenance functionality retrospectively fitted. Is this a blocker for its acceptenance?

>3. How applicable will this be to a cloud provider like GCP that performs live migrations of their VMs?

## Summary

This enhancement proposal explores the idea of having a machine-maintenance-operator(MMO) that will inspect each machine CR for scheduled events for each infra/worker machine in the cluster. Upon finding a maintenance for a machine, the MMO will seek to execute this maintenance manually at the earliest convenient time as defined by an administrator. 

A POC of the idea can be found [here](https://github.com/dofinn/machine-maintenance-operator).

## Motivation

> Currently the machine-api marks an machine as stopped when they have been terminated/stopped by the cloud provider. A provider may terminate/stop and instance when there is an associated maintenance scheduled for it. This is also the case for users manually terminating/stopping machines via the console. The MMO is a proactive approach to executing the required (delete target machine CR)to enable the machine-api to manage machinesets. This operator would work in conjunction with MachineHealthCheck implimentation. 

Reviewer: couple grammar nits and i was also having trouble with the last sentence. perhaps this is close to what you meant?

```suggestion
Currently the machine-api marks a machine as stopped when it has been terminated/stopped by the cloud provider. A provider may terminate/stop an instance when there is an associated maintenance scheduled for it. This is also the case for users manually terminating/stopping machines via the console. The MMO is a proactive approach to executing the required action (for example, delete target machine CR) to enable the machine-api to manage machinesets. This operator would work in conjunction with MachineHealthCheck implementation. 
```

## Requirements

> That the machine-api collects schedules maintenaces from cloud providers and posts them in the status of the each machines CR

Reviewer: should this be "scheduled maintenances" ?

```suggestion
That the machine-api collects scheduled maintenances from cloud providers and posts them in the status of each machine's CR
```

### Goals

List the specific goals of the proposal. How will we know that this has succeeded?
The machine-maintenance-operator will:
* detect maintenances from machine CRs and execute them manually at their earliest and most convenient time

### Non-Goals

This is not a solution for managing maintenances or machine state for master machine roles. 

## Proposal

### User Stories [optional]

#### Story 1
As a customer using OKD/OCP/OSD, I want my cluster/s to be proactive in handling events that are initiated by the cloud provider hosting my cluster. This will assure me that OKD/OCP/OSD accounts for machines in the cluster that require terminating or stopping to be proactively maintained and addressed with intention. 

### Implementation Details/Notes/Constraints [optional]

Constraints: 

> * This implimentation will require the machine-api to query cloud providers for scheduled maintenances and publish them in the machines CR. 

Reviewer: minor nits

```suggestion
* This implementation will require the machine-api to query cloud providers for scheduled maintenances and publish them in the machine's CR. 
```

> * This implementation will require the machine-api to query cloud providers for scheduled maintenances and publish them in the machine's CR. 

Reviewer 2: Seems like the current proposal records these in `MachineMaintenance` resources, instead of writing to the `Machine` resource.  Makes sense to me, because you don't want to be fighting the machine-API trying to write to the same `Machine` resource, but probably need to update this line to talk about "the machine-maintenance resource".

Author: Thanks. Resolved: c29b1c0f1c8f33616045c38793009c6e44262acc

* GCP only allows maintenances to be queried from the node itself -> `curl http://metadata.google.internal/computeMetadata/v1/instance/maintenance-event -H "Metadata-Flavor: Google"`

> This operator will iterate through the machineList{} and inspect each machine CR for scheduled maintenances. If a maintenance is found, the controller will validate the state of the cluster prior to performing any maintenance. For example; is the cluster upgrading? is the cluster already performing a maintenance?

Reviewer 2: "is the cluster upgrading" should be orthogonal.  The limiting conditions are "are we already attempting to cordon/drain another machine in this pool?" and "can we cordon and drain this machine?".  Workloads on the machines should be protected by PDBs and such to prevent eviction that would impact cluster health or provided services.

Author: Thanks. Resolved: 4556f9ed701b21aae453bf67f0a4aa33d908a57f

These processes will only hold true for infra and worker roles of machines within the cluster. 

> If a scheduled maintenance is detected for a master node, an alert should be raised.

Reviewer 2: Why is this?  Does scheduled maintenance wipe the current disk or something?

Author: Admins should be aware that a provider maintenance is planned for a control plane node?

> Not every cloud provider has the same kind of maintenances that inherit that same degradation on the cluster.  Some of the features may need to be toggleable per provider. 

Reviewer: i'm not quite following the first sentence. is this saying that individual maintenance features would need to be toggleable, or something else?

Author: yes. For example, we may not want the machine maintenance operator to do anything on a GCP cluster due to the nature of GCP VM migrations. Please see https://cloud.google.com/compute/docs/storing-retrieving-metadata#maintenanceevents

Reviewer 2: Can we spell this out in the enhancement text?

> Example: (AWS maintenance)[https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/monitoring-instances-status-check_sched.html#types-of-scheduled-events] vs (GCP maintenance)[https://cloud.google.com/compute/docs/storing-retrieving-metadata#maintenanceevents].

Reviewer 2: Please explain in the enhancement text the points that readers are expected to take from this comparison.

Author: Thanks. Resolved: c760ac9aad824b98bf380e55d9855d8c6a092e1b

### Risks and Mitigations

## Design Details

AWS event types:
* instance-stop
* instance-retirement
* instance-reboot
* system-reboot
* system-maintenance

### machinemaintenance controller

> The machinemaintenance controller will iterate through machine CRs and reconcile identified mainteances. It will be responsible for first validating the state of the cluster prior to executing anything on a target object. Validating the state of the cluster will include will initially check for only is the cluster upgrading or is a maintenance already being performed. More use-cases can be added as seen fit. 

Reviewer: spelling

```suggestion
The machinemaintenance controller will iterate through machine CRs and reconcile identified maintenances. It will be responsible for first validating the state of the cluster prior to executing anything on a target object. Validating the state of the cluster will include will initially check for only is the cluster upgrading or is a maintenance already being performed. More use-cases can be added as seen fit. 
```

> If the cluster fails validation (for example is upgrading), the controller will requeue the object and process it again according to its `SyncPeriod` which would currently be proposed at 60 minutes. 

Reviewer 2: I think "60 minutes" is overly-specific for an enhancement proposal.  Can we just say "... process it again later." and leave the duration as an internal implementation detail?

Author: Thanks. Resolved: 4fb294419dbf2a90a89c7d58a7890ab3f0a7b806

> After cluster validation, the controller will ascertain if its in a maintenance window where is can execute maintenances (See open question 2). If in the case no maintenance windows are defined, the controller will continue as true. If the maintenance window logic is only applicable in OSD, the operator could validate if its "managed" prior to expecting these resources.

Reviewer: minor grammar nit

```suggestion
After cluster validation, the controller will ascertain if its in a maintenance window where it can execute maintenances (See open question 2). In the case that no maintenance windows are defined, the controller will continue as true. If the maintenance window logic is only applicable in OSD, the operator could validate if its "managed" prior to expecting these resources.
```

The event type is then sourced from the CR and then is resolved by either deleting a target machine CR (so the machine-api creates a new one) or raising an alert for manual intervention (master maintenance scheduled). 

> This is a very UDP type of approach. 

Reviewer 2: It's not clear to me what this comparison adds.  If you want to talk about a particular feature of this approach, can you talk about it directly in the enhancement text, instead of referencing another protocol and leaving it to the reader to look for parallels.

Author: Removed protocol reference.

> The MMo could also look to store state of its actions in its on machinemaintenance CR that it would create from a machine CR. 

Reviewer: minor grammar nit

```suggestion
The MMo could also look to store the state of its actions in its own machinemaintenance CR that it would create from a machine CR. 
```

### Example machinemaintenance CR

```
apiVersion: machinemaintenance.managed.openshift.io/v1alpha1
kind: MachineMaintenance
metadata:
  name: mm-dofinn-20201705-blz22-worker-ap-southeast-2a-984h4
  namespace: machine-maintenance-operator
spec:
  maintenancescheduled: false
  eventcode: "instance-stop"
  eventid: "instance-event-01d0903276a5d038c"
  machineid: "i-04fdc4494938e5c4a"
  machinelink: "dofinn-20201705-blz22-worker-ap-southeast-2a-984h4"
status:
  maintenance: "in-progress"
```

> The MMO would then delete this CR after validating the target machine has been deleted and new one created indicating that the maintenance is completed. 

Reviewer: would there be any way to know that the maintenance had completed after the CR is deleted?

my concern is that the maintenance would complete and then the deletion happens and due to timing something might miss that signal. in these cases is there some backup or audit trail (perhaps events) ?

Author: >would there be any way to know that the maintenance had completed after the CR is deleted?
Make the exact same query to find any scheduled events for target node. If != nil, either the maintenance didnt clear or a new one is now there.

### Alerting
An alert will be raised for the following conditions:

* Unable to schedule maintenance during customer defined window and prior to cloud provider maintenance deadline

> * Post maintenance verification failed

Reviewer 2: Do you define post-maintenance verification in the enhancement?  If so, I'm having trouble finding it.

Author: https://github.com/openshift/enhancements/pull/341/files#diff-9576061bff3e7a1922fa6c3ab7de01daR130

* Node does not return after shutdown
* Node unable to drain

> * Node unable to shutdown

Reviewer 2: I would expect drain/shutdown to be MachineHealthCheck / machine-API operator concerns that a machine-maintenance operator would not need to handle directly.

Author: MMO would only alert if it was the process that triggered the drain. Nodes being unable to perform drain/shutdown would require and alert for manual intervention.

## Open Questions
>1. How applicable is this use-case to baremetal installations? 

>2. The ability to declare maintenance windows is currently being discussed in OSD. Nothing has be design in concrete meaning that this proposal has an undesigned depedency. However develop could commence on this operator and have the maintenance functionality retrospectively fitted. Is this a blocker for its acceptenance?

>3. How applicable will this be to a cloud provider like GCP that performs live migrations of their VMs?

## Summary

This enhancement proposal explores the idea of having a machine-maintenance-operator(MMO) that acts as a machine 'nanny'. The MMO will actively search for scheduled maintenances for each infra/worker machine in the cluster along with validating the state of these machines against the cloud providers API. Upon finding a maintenance for a machine, the MMO will seek to execute this maintenance manually at the earliest convenient time as defined by an administrator. 

A POC of the idea can be found [here](https://github.com/dofinn/machine-maintenance-operator).

## Motivation

> Currently the machine-api fails to accurately detect the state of machines (validated in AWS) when they have been terminated/stopped by the cloud provider. A provider may terminate/stop and instance when there is an associated maintenance scheduled for it. This is also the case for users manually terminating/stopping machines via the console (validated in AWS). The MMO is a proactive approach to first detecting maintenances and/or machine state from the cloud provider, then executing the required to enable the machine-api to continue managing machines and machinesets. This operator would work in conjunction with MachineHealthCheck implimentation. 

Reviewer 3: This is untrue.  We track the instance state every 15 minutes, and it should be available in the status field.  This is often confused with machine phase, which may be "Running" even if the instance is stopped.  As mentioned elsewhere, the word "Running" is not a great fit, what it really means is "MachineBecameANode" and then that phase is unchanged until the machine is deleted.

Author: I stand corrected. 

```
oc get machine gxs-aws-nj7hf-worker-ap-southeast-2a-mmjwc -o json | jq '.status.providerStatus.instanceState'
"running"
```

```
oc get machine gxs-aws-nj7hf-worker-ap-southeast-2a-mmjwc -o json | jq '.status.providerStatus.instanceState'
"stopping"
```

```
oc get machine gxs-aws-nj7hf-worker-ap-southeast-2a-mmjwc -o json | jq '.status.providerStatus.instanceState'
"stopped"
```

Is there a remediation path from this state? or is the most agnostic solution just to log the current status and act elsewhere?

Reviewer 3: By default, the machine-controller does not do anything.  Eventually, the node will go unready.  Using MachineHealthChecks will cause the machine to be deleted (if it's part of a machineset), and you'll get a new machine.

List the specific goals of the proposal. How will we know that this has succeeded?
The machine-maintenance-operator will:
* detect maintenances and execute them manually at their earliest and most convenient time
* validate the state of the machine against its machine CR and the cloud providers API
* validate the state of the machine behind a cloud providers load balancers and provide an alert upon recognition of a miss-match of state

### Non-Goals

This is not a solution for managing maintenances or machine state for master machine roles. 

## Proposal

### User Stories [optional]

#### Story 1
As a customer using OKD/OCP/OSD, I want my cluster/s to be proactive in handling events that are initiated by the cloud provider hosting my cluster. This will assure me that OKD/OCP/OSD accounts for machines in the cluster that require terminating or stopping to be proactively maintained and addressed with intention. 

#### Story 2
As an SRE supporting OSD, I want automated detection of cloud provider scheduled maintenance events. This will enable automated processes to proactively handle the event and execute the required with intent and validation, while also alerting me if required. 

> As an SRE supporting OSD, I want automated validation of machine state as advertised by the machine-api against the cloud provider. This will enable automated processes to proactively handle the event and execute the required with intent and validation, while also alerting me if required. 

Reviewer 3: Not clear what it is you want here.  What is validation of machine state?

Author: To maintain a set of machines whoms `.status.providerStatus.instanceState` = "running" equal with that of the number of machines in the machineset.

> As an SRE supporting OSD, I want automated validation of machine state as advertised by the machine-api against the cloud providers load balancers (health checks). This will enable automated processes to proactively handle the event and execute the required with intent and validation, while also alerting me if required. 

Reviewer 3: It's not clear what load balancers really have to do with machine state.  ELBs have health checks.  If a machine fails health checks, it's removed from ELBs.

For other types of load balancers (network ingress?), those are handled by operators, and again, health checks.

Author: This operator would alert on (potentially remediate) the removal of a node from an ELB on specific conditions. For example a failed health check.

Reviewer 3: I guess I need to know more about how instances are ending up behind the ELBs in the first place.  I thought the network ingress operator took care of this.

If a scheduled maintenance is detected for a master node, an alert should be raised.

Not every cloud provider has the same kind of maintenances that inherit that same degradation on the cluster.  Some of the features may need to be toggleable per provider. 

Example: (AWS maintenance)[https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/monitoring-instances-status-check_sched.html#types-of-scheduled-events] vs (GCP maintenance)[https://cloud.google.com/compute/docs/storing-retrieving-metadata#maintenanceevents].

### Risks and Mitigations

## Design Details

### maintenanceWatcher
The maintenanceWatcher is a go routine that currently cycles every 60 minutes. It retrieves a list of machines and iterates through them to check if each machine has an associated scheduled maintenance. If a maintenance is found for a machine, a machinemaintenance CR is created with the event type from that cloud provider for reconciliation by the machinemaintenance controller. 

For example, AWS events will result in the event type being either:
* instance-stop
* instance-retirement
* instance-reboot
* system-reboot
* system-maintenance

### stateWatcher

> The stateWatcher is a go routine that currently cycles every 10 minutes. It retrieves a list of machines and iterates through them to validate the state of machine from the machine-api with that of the cloud provider. If miss-match in state is found, machinemaintenance CR is created with event type "machine-state-missmatch" for reconciliation by the machinemaintenance controller.

Reviewer 3: It's not clear what, if any, action is going to be taken after this is noticed.  In any case, as noted above, this is wholly covered by the machine-api already.

### lbWatcher

> The lbWatcher is a go routine that currently cycles every 10 minutes. It retrieves a list of machines and iterates through them to validate the health of machine from the machine-api with that of the cloud providers load balancers. This will only be targeted at infra and master machines. A machinemaintenance CR will be created with event type "lb-unhealthy" if a host is unexpectedly found unhealthy.

Reviewer 3: Most 4.x clusters don't have dedicated infra machines.  As noted above, this is handled by a load balancer's health checks and whatever operator is deploying a particular pod that needs a service load balancer.
